ggml.h

```diff
@@ -371,10 +373,12 @@ extern "C" {
         int64_t perf_time_us;
 
         void * data;
+        void * data_device[GGML_MAX_DEVICES];
+        int i_device;
```

Comment:
Consider moving all GPU related members into a struct in `ggml-cuda.h` which is referenced via:

```c
struct ggml_tensor {
    ...
    void * data_extra; // user-managed
    ...
};
```

For example:

```c
// ggml-cuda.h

struct ggml_tensor_gpu {
    int layer;
    void * device[GGML_CUDA_MAX_DEVICES];
    int idev;
};

// llama.cpp / ggml-cuda.cpp

...
struct ggml_tensor_gpu my_gpu_data;
tensor->data_extra = &my_gpu_data;
...
```

The extra structs will be managed by the user-code (i.e. `llama.cpp`) or by `ggml-cuda` / `ggml-opencl`.
The goal is to avoid coupling `ggml` core structs with LLM- / GPU- specific things, so that the interface remains more general-purpose.

Reply:
What if NUMA nodes are considered different devices too?

Reply:
Should I put the int representing the layer also into the GPU extra data struct or should it stay in `ggml_tensor`?

Reply:
I decided to keep `layer` in `ggml_tensor`. Otherwise you would need to start adding extra data to `ggml_tensor` outside of `ggml-cuda.cu` (or extract it from the name).

Reply:
> What if NUMA nodes are considered different devices too?

Unlikely

> I decided to keep layer in ggml_tensor

The `layer` is specific to deep neural nets and it is not used in `ggml.c` as far as I can tell.
I think it should be moved out of `ggml_tensor`

Reply:
I  also added a define `GGML_MAX_DEVICES` and a utility function `ggml_nbytes_split` to `ggml.h`. Should I move those as well?

Reply:
The `GGML_MAX_DEVICES` - yes

The `ggml_nbytes_split()` can go either way - as you prefer.